# **Найти лучшую модель**

In [1]:
pip install lazypredict > /dev/null 2>&1

Note: you may need to restart the kernel to use updated packages.


‘Ёб⥬Ґ ­Ґ г¤ Ґвбп ­ ©вЁ гЄ § ­­л© Їгвм.


In [1]:
pip install lazypredict

Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.

   ---------------------------------------- 0.0/1.3 MB ? eta -:--:--
    --------------------------------------- 0.0/1.3 MB ? eta -:--:--
   -- ------------------------------------- 0.1/1.3 MB 744.7 kB/s eta 0:00:02
   -- ------------------------------------- 0.1/1.3 MB 744.7 kB/s eta 0:00:02
   ---- ----------------------------------- 0.2/1.3 MB 706.2 kB/s eta 0:00:02
   ---- ----------------------------------- 0.2/1.3 MB 706.2 kB/s eta 0:00:02
   ----- ---------------------------------- 0.2/1.3 MB 523.5 kB/s eta 0:00:03
   -------- ------------------------------- 0.3/1.3 MB 811.7 kB/s eta 0:00:02
   -------- ------------------------------- 0.3/1.3 MB 811.7 kB/s eta 0:00:02
   -------- ------------------------------- 0.3/1.3 MB 707.1 kB/s eta 0:00:02
   -------- ------------------------------- 0.3/1.3 MB 707.1 kB/s eta 0:00:02
   -------------

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


## **Imports**

In [2]:
import pandas as pd
import numpy as np
import gc
from sklearn.model_selection import train_test_split
from lazypredict.Supervised import LazyClassifier
from keras.models import  Sequential
from sklearn.preprocessing import LabelEncoder
import random


## **Import data**

### Эти данные могут быть использованы для обнаружения трафика троянских программ. Данные содержат записи о трафике, таком как троянские программы и вредоносные программы, поэтому обнаружение троянских программ и доброкачественных программ может быть выполнено с использованием двоичной классификации. Данные получены с веб-сайта CIC. В наборе данных записано следующее количество образцов:

### Троянские данные с 90 683 наблюдениями
### Данные о вредоносных программах с 86 799 наблюдениями

In [3]:
data = pd.read_csv('Trojan_Detection.csv')

# Сохраняйем 10 тысяч строк, чтобы найти лучшую модель (во избежание перегрузки памяти)
data = data.iloc[:10000]

print(data.shape)

(10000, 86)


## **Столбцы кодировки**

In [4]:
lab=LabelEncoder()

# Кодирование функций
for i  in data.select_dtypes(include="object").columns.values:
    data[i]=lab.fit_transform(data[i])

# Кодирование меток
data['Class']=lab.fit_transform(data['Class'])

# Разделение данных

In [5]:
x = data.iloc[:, 0:-2]
y = data.iloc[:, -1]

In [6]:
x_train,x_test,y_train,y_test=train_test_split(x, y, test_size=0.20, train_size=0.80)

## Lazy Classifier
### LazyClassifier из пакета lazypredict для автоматического поиска и сравнения различных моделей машинного обучения без необходимости явного указания каждой модели отдельно.

In [7]:
lazy=LazyClassifier(verbose=1, ignore_warnings=False, predictions=True)
models,predict=lazy.fit(x_train,x_test,y_train,y_test)
print(models)

  3%|██▊                                                                                | 1/29 [00:00<00:17,  1.57it/s]

{'Model': 'AdaBoostClassifier', 'Accuracy': 1.0, 'Balanced Accuracy': 1.0, 'ROC AUC': 1.0, 'F1 Score': 1.0, 'Time taken': 0.6364600658416748}


  7%|█████▋                                                                             | 2/29 [00:01<00:18,  1.43it/s]

{'Model': 'BaggingClassifier', 'Accuracy': 1.0, 'Balanced Accuracy': 1.0, 'ROC AUC': 1.0, 'F1 Score': 1.0, 'Time taken': 0.7473862171173096}


 10%|████████▌                                                                          | 3/29 [00:01<00:14,  1.74it/s]

{'Model': 'BernoulliNB', 'Accuracy': 0.98, 'Balanced Accuracy': 0.9799404270231034, 'ROC AUC': 0.9799404270231032, 'F1 Score': 0.9799992790656692, 'Time taken': 0.42410898208618164}


 14%|███████████▍                                                                       | 4/29 [00:03<00:22,  1.10it/s]

{'Model': 'CalibratedClassifierCV', 'Accuracy': 0.999, 'Balanced Accuracy': 0.9990338164251208, 'ROC AUC': 0.9990338164251208, 'F1 Score': 0.9990000340393494, 'Time taken': 1.4244163036346436}
CategoricalNB model failed to execute
Negative values in data passed to CategoricalNB (input X)


 24%|████████████████████                                                               | 7/29 [00:03<00:08,  2.56it/s]

{'Model': 'DecisionTreeClassifier', 'Accuracy': 1.0, 'Balanced Accuracy': 1.0, 'ROC AUC': 1.0, 'F1 Score': 1.0, 'Time taken': 0.14184904098510742}
{'Model': 'DummyClassifier', 'Accuracy': 0.5175, 'Balanced Accuracy': 0.5, 'ROC AUC': 0.5, 'F1 Score': 0.3529571663920922, 'Time taken': 0.12479257583618164}


 28%|██████████████████████▉                                                            | 8/29 [00:03<00:06,  3.09it/s]

{'Model': 'ExtraTreeClassifier', 'Accuracy': 0.9375, 'Balanced Accuracy': 0.9376160796976296, 'ROC AUC': 0.9376160796976296, 'F1 Score': 0.9375101494575655, 'Time taken': 0.15070152282714844}


 34%|████████████████████████████▎                                                     | 10/29 [00:05<00:10,  1.76it/s]

{'Model': 'ExtraTreesClassifier', 'Accuracy': 0.993, 'Balanced Accuracy': 0.9928512427723962, 'ROC AUC': 0.9928512427723963, 'F1 Score': 0.9929989063365379, 'Time taken': 1.7845005989074707}
{'Model': 'GaussianNB', 'Accuracy': 0.914, 'Balanced Accuracy': 0.9130885334534805, 'ROC AUC': 0.9130885334534805, 'F1 Score': 0.9138848887658081, 'Time taken': 0.13766908645629883}


 38%|███████████████████████████████                                                   | 11/29 [00:06<00:10,  1.68it/s]

{'Model': 'KNeighborsClassifier', 'Accuracy': 0.939, 'Balanced Accuracy': 0.9387149257840854, 'ROC AUC': 0.9387149257840854, 'F1 Score': 0.9389849687324393, 'Time taken': 0.6566791534423828}


 41%|█████████████████████████████████▉                                                | 12/29 [00:16<00:56,  3.31s/it]

{'Model': 'LabelPropagation', 'Accuracy': 0.932, 'Balanced Accuracy': 0.9323020700357938, 'ROC AUC': 0.9323020700357939, 'F1 Score': 0.9320179623463115, 'Time taken': 9.782914400100708}


 45%|████████████████████████████████████▊                                             | 13/29 [00:28<01:35,  6.00s/it]

{'Model': 'LabelSpreading', 'Accuracy': 0.932, 'Balanced Accuracy': 0.9323020700357938, 'ROC AUC': 0.9323020700357939, 'F1 Score': 0.9320179623463115, 'Time taken': 12.36513352394104}


 48%|███████████████████████████████████████▌                                          | 14/29 [00:29<01:09,  4.64s/it]

{'Model': 'LinearDiscriminantAnalysis', 'Accuracy': 0.953, 'Balanced Accuracy': 0.952276538759981, 'ROC AUC': 0.9522765387599811, 'F1 Score': 0.9529516369734145, 'Time taken': 1.4505043029785156}


 52%|██████████████████████████████████████████▍                                       | 15/29 [00:30<00:46,  3.36s/it]

{'Model': 'LinearSVC', 'Accuracy': 0.9995, 'Balanced Accuracy': 0.9995169082125603, 'ROC AUC': 0.9995169082125603, 'F1 Score': 0.9995000086352782, 'Time taken': 0.3319835662841797}


 59%|████████████████████████████████████████████████                                  | 17/29 [00:31<00:22,  1.89s/it]

{'Model': 'LogisticRegression', 'Accuracy': 0.9975, 'Balanced Accuracy': 0.9974443693524568, 'ROC AUC': 0.9974443693524567, 'F1 Score': 0.9974998629424051, 'Time taken': 0.9127142429351807}
{'Model': 'NearestCentroid', 'Accuracy': 0.9445, 'Balanced Accuracy': 0.9435032915321269, 'ROC AUC': 0.9435032915321269, 'F1 Score': 0.944409205467379, 'Time taken': 0.16081571578979492}


 62%|██████████████████████████████████████████████████▉                               | 18/29 [00:49<01:13,  6.72s/it]

{'Model': 'NuSVC', 'Accuracy': 0.9225, 'Balanced Accuracy': 0.9205301494330556, 'ROC AUC': 0.9205301494330556, 'F1 Score': 0.9221301601972045, 'Time taken': 18.039297103881836}


 66%|█████████████████████████████████████████████████████▋                            | 19/29 [00:49<00:48,  4.81s/it]

{'Model': 'PassiveAggressiveClassifier', 'Accuracy': 0.997, 'Balanced Accuracy': 0.9969262346374308, 'ROC AUC': 0.9969262346374308, 'F1 Score': 0.9969997776956654, 'Time taken': 0.32238221168518066}


 69%|████████████████████████████████████████████████████████▌                         | 20/29 [00:49<00:30,  3.44s/it]

{'Model': 'Perceptron', 'Accuracy': 0.9855, 'Balanced Accuracy': 0.9854646942504568, 'ROC AUC': 0.9854646942504568, 'F1 Score': 0.9854997423002338, 'Time taken': 0.2262897491455078}


 72%|███████████████████████████████████████████████████████████▍                      | 21/29 [00:50<00:19,  2.48s/it]

{'Model': 'QuadraticDiscriminantAnalysis', 'Accuracy': 0.834, 'Balanced Accuracy': 0.8352331606217617, 'ROC AUC': 0.8352331606217617, 'F1 Score': 0.833959162529851, 'Time taken': 0.23348426818847656}


 76%|██████████████████████████████████████████████████████████████▏                   | 22/29 [00:53<00:19,  2.75s/it]

{'Model': 'RandomForestClassifier', 'Accuracy': 1.0, 'Balanced Accuracy': 1.0, 'ROC AUC': 1.0, 'F1 Score': 1.0, 'Time taken': 3.3935658931732178}


 79%|█████████████████████████████████████████████████████████████████                 | 23/29 [00:54<00:12,  2.06s/it]

{'Model': 'RidgeClassifier', 'Accuracy': 0.953, 'Balanced Accuracy': 0.952276538759981, 'ROC AUC': 0.9522765387599811, 'F1 Score': 0.9529516369734145, 'Time taken': 0.44803595542907715}


 86%|██████████████████████████████████████████████████████████████████████▋           | 25/29 [00:54<00:04,  1.12s/it]

{'Model': 'RidgeClassifierCV', 'Accuracy': 0.953, 'Balanced Accuracy': 0.952276538759981, 'ROC AUC': 0.9522765387599811, 'F1 Score': 0.9529516369734145, 'Time taken': 0.24934101104736328}
{'Model': 'SGDClassifier', 'Accuracy': 0.993, 'Balanced Accuracy': 0.992956371555155, 'ROC AUC': 0.9929563715551549, 'F1 Score': 0.9929997476729843, 'Time taken': 0.20196843147277832}


 90%|█████████████████████████████████████████████████████████████████████████▌        | 26/29 [00:59<00:06,  2.31s/it]

{'Model': 'SVC', 'Accuracy': 0.96, 'Balanced Accuracy': 0.959215038422067, 'ROC AUC': 0.959215038422067, 'F1 Score': 0.95995281201849, 'Time taken': 5.066623687744141}
StackingClassifier model failed to execute
StackingClassifier.__init__() missing 1 required positional argument: 'estimators'


 97%|███████████████████████████████████████████████████████████████████████████████▏  | 28/29 [01:00<00:01,  1.47s/it]

{'Model': 'XGBClassifier', 'Accuracy': 1.0, 'Balanced Accuracy': 1.0, 'ROC AUC': 1.0, 'F1 Score': 1.0, 'Time taken': 0.9879395961761475}
[LightGBM] [Info] Number of positive: 4118, number of negative: 3882
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.005426 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 14277
[LightGBM] [Info] Number of data points in the train set: 8000, number of used features: 72
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.514750 -> initscore=0.059017
[LightGBM] [Info] Start training from score 0.059017
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -in

100%|██████████████████████████████████████████████████████████████████████████████████| 29/29 [01:04<00:00,  2.08s/it]

[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

100%|██████████████████████████████████████████████████████████████████████████████████| 29/29 [01:04<00:00,  2.22s/it]

                               Accuracy  Balanced Accuracy  ROC AUC  F1 Score  \
Model                                                                           
AdaBoostClassifier                 1.00               1.00     1.00      1.00   
BaggingClassifier                  1.00               1.00     1.00      1.00   
XGBClassifier                      1.00               1.00     1.00      1.00   
RandomForestClassifier             1.00               1.00     1.00      1.00   
LGBMClassifier                     1.00               1.00     1.00      1.00   
DecisionTreeClassifier             1.00               1.00     1.00      1.00   
LinearSVC                          1.00               1.00     1.00      1.00   
CalibratedClassifierCV             1.00               1.00     1.00      1.00   
LogisticRegression                 1.00               1.00     1.00      1.00   
PassiveAggressiveClassifier        1.00               1.00     1.00      1.00   
SGDClassifier               

## Протестируем лучшую модель

## **Импорт**

In [8]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

## **Импорт данных**

In [9]:
data = pd.read_csv('/kaggle/input/trojan-detection/Trojan_Detection.csv')

print(data.shape)

(177482, 86)


## **Столбцы кодировки**

In [10]:
lab=LabelEncoder()

# Кодирование функций
for i  in data.select_dtypes(include="object").columns.values:
    data[i]=lab.fit_transform(data[i])

# Кодирование меток
data['Class']=lab.fit_transform(data['Class'])

## **Разделение данных**

In [11]:
x = data.iloc[:, 0:-2]
y = data.iloc[:, -1]

In [12]:
x_train,x_test,y_train,y_test=train_test_split(x, y, test_size=0.20, train_size=0.80)

In [13]:
# # Инициализировать классификатор AdaBoost
model = AdaBoostClassifier(n_estimators=50, learning_rate=1, random_state=42)

# Подогнать модель под данные обучения
model.fit(x_train, y_train)

AdaBoostClassifier(learning_rate=1, random_state=42)

## **Тестовая модель на тестовом наборе**

In [14]:
# Предсказать результаты тестового набора
y_pred  = model.predict(x_test)

# Calculate the accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print(f'Confusion Matrix for AdaBoostClassifier:\n{conf_matrix}')

# Отчет о классификации
class_report = classification_report(y_test, y_pred)
print(f'Classification Report for AdaBoostClassifier:\n{class_report}')

Accuracy: 1.00
Confusion Matrix for AdaBoostClassifier:
[[17367     1]
 [    0 18129]]
Classification Report for AdaBoostClassifier:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     17368
           1       1.00      1.00      1.00     18129

    accuracy                           1.00     35497
   macro avg       1.00      1.00      1.00     35497
weighted avg       1.00      1.00      1.00     35497



#### Результаты оценки модели AdaBoostClassifier очень впечатляющие:

Accuracy (точность): Значение 1.00 означает, что модель правильно классифицировала все примеры в тестовом наборе данных, что свидетельствует о высоком качестве классификации.

Confusion Matrix (матрица ошибок): Матрица показывает, что модель правильно предсказала 17367 негативных примеров (true negatives) и 18129 положительных примеров (true positives), при этом совершив всего одну ошибку false negative и ни одной ошибки false positive. Это означает, что модель неверно классифицировала только один положительный пример из 35497.

Classification Report (отчет о классификации): Отчет также подтверждает идеальное качество классификации для обоих классов (0 и 1) с высокими значениями precision, recall и f1-score, равными 1.00. Также отметим, что значение поддержки (support) для каждого класса соответствует реальному количеству примеров в тестовом наборе данных.

В целом, эти результаты говорят о том, что модель AdaBoostClassifier успешно обучилась на обучающих данных и хорошо обобщила полученные знания на тестовых данных, обеспечивая идеальное качество классификации.